src/FSharp.Data.GraphQL.Server.AppInfrastructure/GraphQLWebsocketMiddleware.fs

```diff
@@ -0,0 +1,375 @@
+namespace FSharp.Data.GraphQL.Server.AppInfrastructure
+
+open FSharp.Data.GraphQL
+open Microsoft.AspNetCore.Http
+open FSharp.Data.GraphQL.Server.AppInfrastructure.Rop
+open System
+open System.Collections.Generic
+open System.Net.WebSockets
+open System.Text.Json
+open System.Threading
+open System.Threading.Tasks
+open FSharp.Data.GraphQL.Execution
+open Microsoft.Extensions.Hosting
+open Microsoft.Extensions.Logging
+
+type GraphQLWebSocketMiddleware<'Root>(next : RequestDelegate, applicationLifetime : IHostApplicationLifetime, serviceProvider : IServiceProvider, logger : ILogger<GraphQLWebSocketMiddleware<'Root>>, options : GraphQLOptions<'Root>) =
+
+  let serializeServerMessage (jsonSerializerOptions: JsonSerializerOptions) (serverMessage : ServerMessage) =
+      task {
+        let raw =
+          match serverMessage with
+          | ConnectionAck ->
+            { Id = None
+              Type = "connection_ack"
+              Payload = None }
+          | ServerPing ->
+            { Id = None
+              Type = "ping"
+              Payload = None }
+          | ServerPong p ->
+            { Id = None
+              Type = "pong"
+              Payload = p |> Option.map CustomResponse }
+          | Next (id, payload) ->
+            { Id = Some id
+              Type = "next"
+              Payload = Some <| ExecutionResult payload }
+          | Complete id ->
+            { Id = Some id
+              Type = "complete"
+              Payload = None }
+          | Error (id, errMsgs) ->
+            { Id = Some id
+              Type = "error"
+              Payload = Some <| ErrorMessages errMsgs }
+        return JsonSerializer.Serialize(raw, jsonSerializerOptions)
+      }
+
+  let deserializeClientMessage (serializerOptions : JsonSerializerOptions) (msg: string) =
+    task {
+        try
+          return
+            JsonSerializer.Deserialize<ClientMessage>(msg, serializerOptions)
+            |> succeed
+        with
+        | :? InvalidMessageException as e ->
+          return
+            fail <| InvalidMessage(4400, e.Message.ToString())
+        | :? JsonException as e ->
+          if logger.IsEnabled(LogLevel.Debug) then
+            logger.LogDebug(e.ToString())
+          else
+            ()
+          return
+            fail <| InvalidMessage(4400, "invalid json in client message")
+    }
+
+  let isSocketOpen (theSocket : WebSocket) =
+    not (theSocket.State = WebSocketState.Aborted) &&
+    not (theSocket.State = WebSocketState.Closed) &&
+    not (theSocket.State = WebSocketState.CloseReceived)
+
+  let canCloseSocket (theSocket : WebSocket) =
+    not (theSocket.State = WebSocketState.Aborted) &&
+    not (theSocket.State = WebSocketState.Closed)
+
+  let receiveMessageViaSocket (cancellationToken : CancellationToken) (serializerOptions: JsonSerializerOptions) (executor : Executor<'Root>) (socket : WebSocket) : Task<RopResult<ClientMessage, ClientMessageProtocolFailure> option> =
```

Comment:
This function is receiving a lot of arguments. Unless we need to use them in partial application, maybe we could group them in a single object, except for the cancellation token? An `Context` type, maybe?

Reply:
It's good that you pointed it out, now I just saw that the `executor` parameter is not necessary. It is a remnant of a previous approach (now the executor that was needed here is needed in the JsonConverters).
Without the `executor` parameter, there remain only three parameters (instead of four):
1. the cancellation token;
2. the serializer options;
3. the socket;

See a usage example in the function `safe_Receive()` (which abstracts away the `cancellationToken` parameter by calling `safe_ReceiveMessageViaSocket`):

```fsharp
    let safe_Receive() =
      socket
      |> safe_ReceiveMessageViaSocket serializerOptions
```

I find that pretty ok for a number of parameters, if you ask me.